### PR TITLE
QOL changes in Basic to replace asString by CustomStringConvertible

### PR DESCRIPTION
**Pull request by hartbit**
**TL;DR:** This pull request removes `asString` from SwiftPM, and replaces it with conformances to `CustomStringConvertible` and brings related Quality of Life changes.

This is a pre-requisite PR to a future PR that will introduce a new protocol that extends `CustomStringConvertible`. For that purpose, I need several types (specifically `AbstractPath` and `RelativePath`) to conform to it and return the undecorated path. This change has the benefit of improving the ergonomics of using those types, especially in string interpolation. I also took the opportunity to bring related Quality of Life changes surrounding `CustomStringConvertible` when working with `JSONSerializable` and `OutputByteStream`.

Personally, `asString` always felt weird to me compared to `CustomStringConvertible`, which is the canonical Swift way to transform a value into a string representation. 

### `ByteString`

- Rename `asString` to `validDescription`
- Rename `asReadableString` to `cString`
- Change the implementation of `description` to return `validDescription` and trap on invalid UTF-8 sequences.

Those changes simplify the majority of the code in SwiftPM which attempts to transform `ByteString` into a string by replacing

```swift
byteString.asString!
```
into

```swift
byteString.description
```

To check if the string representation is valid, `validDescription` needs to be used now.

### `AbstractPath` and `RelativePath`

- De-duplicate the `components` implementation by moving it into `PathImpl`
- Remove `asString`
- Change the implementation of `description` to returns the undecorated path

This improves a lot of code in SwiftPM that uses paths in print statement and string interpolation by transforming:

```swift
print(path.asString)
return “Couldn’t read \(path.asString)”
```

into

```swift
print(path)
return “Couldn’t read \(path)”
```

### `JSONSerializable`

Add conditional conformances to `JSONSerializable` to `Array` and `Dictionary` which improves

```swift
return JSON.array(files.map({ $0.toJSON() })
```

into

```swift
return files.toJSON()
```

## `OutputByteStream`

- Add overrides of `<<<` for `CustomStringConvertible` to improve:

```swift
stream <<< file.asString
```

into

```swift
stream <<< file
```

- Add `Format.asJSON` overloads for `CustomStringConvertible` and `[CustomStringConvertible]`, which allows to transform:

```swift
stream <<< Format.asJSON(file.asString)
stream <<< Format.asJSON(files.map({ $0.asString }))
```

into

```swift
stream <<< Format.asJSON(file)
stream <<< Format.asJSON(files)
```

### CODE REVIEW

File: Sources/Basic/ByteString.swift [hunk 1]
```diff
@@ -59,9 +59,21 @@ public struct ByteString: ExpressibleByArrayLiteral, Hashable {
...
+    public var description: String {
+        guard let description = validDescription else {
+            fatalError("invalid byte string: \(cString)")
+        }
+
+        return description
+    }
 
     /// Return the string decoded as a UTF8 sequence, if possible.
-    public var asString: String? {
```
Comment (aciidb0mb3r) [inline Sources/Basic/ByteString.swift]: can we keep `asString` for now and mark them as deprecated?

File: Sources/Basic/Path.swift [hunk 2]
```diff
@@ -338,23 +308,21 @@ extension RelativePath: Codable {
...
-        return lhs.asString < rhs.asString
+        return lhs.description < rhs.description
     }
 }
 
 /// Make absolute paths CustomStringConvertible.
 extension AbsolutePath: CustomStringConvertible {
     public var description: String {
-        // FIXME: We should really be escaping backslashes and quotes here.
-        return "<AbsolutePath:\"\(asString)\">"
```
Comment (aciidb0mb3r) [inline Sources/Basic/Path.swift]: maybe move this to debugDescription

File: Sources/Basic/Path.swift [hunk 3]
```diff
@@ -338,23 +308,21 @@ extension RelativePath: Codable {
...
+        return _impl.string
     }
 }
 
 /// Make relative paths CustomStringConvertible.
 extension RelativePath: CustomStringConvertible {
     public var description: String {
-        // FIXME: We should really be escaping backslashes and quotes here.
-        return "<RelativePath:\"\(asString)\">"
```
Comment (aciidb0mb3r) [inline Sources/Basic/Path.swift]: maybe move this to debugDescription

**Comment by hartbit** [comment]
I finished applying the proposed changes!

**Comment by hartbit** [comment]
@swift-ci please smoke test

**Comment by hartbit** [comment]
@swift-ci please smoke test
